src/coreclr/jit/importer.cpp

```diff
@@ -15989,8 +15989,12 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                         // for the field to operate on the address of the struct.
                         if (!varTypeGCtype(obj->TypeGet()) && impIsValueType(tiObj))
                         {
-                            assert(opcode == CEE_LDFLD && objType != nullptr);
-
+                            if ((opcode != CEE_LDFLD) || (objType == nullptr))
+                            {
+                                BADCODE3("Unexpected opcode (has to be LDFLD) or struct type to access instance field "
```

Comment:
I don't know if this error text surfaces anywhere. If not, what you have is OK.

But if so, it might be better to have two separate checks and two BADCODEs.

The `objType == nullptr` is the case where we don't have a struct type, right? The current error text doesn't spell that out very clearly, and we might want to use ECMA-335 language here. So for the second part, how about something like:

"top of stack was not a value type"

or

"top of stack must be a value type"